src/exec/use_pty/pipe.rs

```diff
@@ -100,16 +100,100 @@ impl<L: Read + Write + AsRawFd, R: Read + Write + AsRawFd> Pipe<L, R> {
+        } else {
+            let &mut Self { start, len, .. } = self;
+            let end = start + len;
+            if end >= self.storage.len() {
+                // Case 1.2. The buffer is not empty and the non-empty section wraps around
+                // `storage`. Meaning that there is only one empty slice in `storage`: `end..start`.
+                let end = end % self.storage.len();
+                read.read(&mut self.storage[end..start])?
+            } else {
+                // Case 1.3. The buffer is non empty and the non-empty section is a contiguous
+                // slice of `storage`. Meaning that there are two empty slices in `storage`:
+                // `..start` and `end..`.
+                let (mid, first_slice) = self.storage.split_at_mut(end);
+                let second_slice = &mut mid[..start];
+                read.read_vectored(&mut [first_slice, second_slice].map(IoSliceMut::new))?
+            }
+        };
+
+        self.len += inserted_len;
+
+        Ok(inserted_len)
+    }
+
+    fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+
+    fn remove<W: Write>(&mut self, write: &mut W) -> io::Result<usize> {
+        let removed_len = if self.is_full() {
+            // Case 2.1. The buffer is full, meaning that there are two non-empty slices in
+            // `storage`: `start..` and `..start`.
+            let (second_slice, first_slice) = self.storage.split_at(self.start);
+            write.write_vectored(&[first_slice, second_slice].map(IoSlice::new))?
+        } else {
+            let end = self.start + self.len;
+            if end >= self.storage.len() {
+                // Case 2.2. The buffer is not full and the non-empty section wraps around
+                // `storage`. Meanning that there are two non-empty slices in `storage`: `start..`
+                // and `..end`.
+                let end = end % self.storage.len();
+                let first_slice = &self.storage[self.start..];
+                let second_slice = &self.storage[..end];
+                write.write_vectored(&[first_slice, second_slice].map(IoSlice::new))?
+            } else {
+                // Case 2.3. The buffer is not full and the non-empty section is a contiguous slice
+                // of `storage.` Meaning that there is only one non-empty slice in `storage`:
+                // `start..end`.
+                write.write(&self.storage[self.start..end])?
+            }
+        };
+
+        self.start += removed_len;
+        self.start %= self.storage.len();
+
+        self.len -= removed_len;
+
+        Ok(removed_len)
+    }
+}
 
 /// A buffer that stores the bytes read from `R` before they are written to `W`.
 struct Buffer<R, W> {
-    internal: [u8; BUFSIZE],
-    /// The start of the busy section of the buffer.
-    start: usize,
-    /// The end of the busy section of the buffer.
-    end: usize,
+    internal: RingBuffer,
```

Comment:
`RingBuffer` should live into a separate module to enforce encapsulation / privacy. right now, this `Buffer` can modify `RingBuffer`'s internal fields (`start`, `len`, etc.) and break `RingBuffer` invariants.